presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushFilterThroughSetOperation.java

```diff
@@ -0,0 +1,110 @@
+
+import static com.facebook.presto.sql.ExpressionUtils.combineConjuncts;
+import static com.facebook.presto.sql.ExpressionUtils.extractConjuncts;
+import static com.facebook.presto.sql.planner.ExpressionSymbolInliner.inlineSymbols;
+import static com.facebook.presto.sql.planner.optimizations.Predicates.predicate;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Collections.emptyList;
+
+public class PushFilterThroughSetOperation
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+    {
+        if (!(node instanceof FilterNode)) {
+            return Optional.empty();
+        }
+
+        FilterNode parent = (FilterNode) node;
+
+        PlanNode source = lookup.resolve(parent.getSource());
+
+        if (!(source instanceof SetOperationNode)) {
+            return Optional.empty();
+        }
+
+        List<Expression> conjuncts = extractConjuncts(parent.getPredicate());
+        SetOperationNode child = (SetOperationNode) source;
+
+        List<Expression> pushableConjuncts = pushableConjuncts(child, conjuncts);
+
+        if (pushableConjuncts.isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<Expression> unpushableConjuncts = unpushableConjuncts(child, conjuncts);
+
+        ImmutableList.Builder<PlanNode> builder = ImmutableList.builder();
+        for (int i = 0; i < child.getSources().size(); i++) {
+            Map<Symbol, SymbolReference> mappings = child.sourceSymbolMap(i);
+            List<Expression> rewrittenConjuncts = pushableConjuncts.stream()
+                    .map(conjunct -> inlineSymbols(mappings, conjunct))
+                    .collect(toImmutableList());
+
+            builder.add(new FilterNode(idAllocator.getNextId(), child.getSources().get(i), combineConjuncts(rewrittenConjuncts)));
+        }
+
+        PlanNode rewrittenChild = child.replaceChildren(builder.build());
+        if (unpushableConjuncts.isEmpty()) {
+            return Optional.of(rewrittenChild);
+        }
+        else {
+            return Optional.of(new FilterNode(idAllocator.getNextId(), rewrittenChild, combineConjuncts(unpushableConjuncts)));
+        }
+    }
+
+    private List<Expression> pushableConjuncts(SetOperationNode child, List<Expression> conjuncts)
+    {
+        if (child instanceof UnionNode) {
+            return conjuncts;
+        }
+
+        return conjuncts.stream()
+                .filter(DeterminismEvaluator::isDeterministic)
+                .collect(toImmutableList());
+    }
+
+    private List<Expression> unpushableConjuncts(SetOperationNode child, List<Expression> conjuncts)
+    {
+        if (child instanceof UnionNode) {
+            return emptyList();
+        }
+
+        return conjuncts.stream()
+                .filter(predicate(DeterminismEvaluator::isDeterministic).negate())
```

Comment:
It looks so. Legacy `PredicatePushDown` does not handle expression determinism either for `Exchange` or `Union`. Though I am not sure if we an assume that planner guarantees this.

@martint What do you think?